### ISSUE TITLE
Installing gd on buster results in freetype-config not found

### ISSUE THREAD

**Issue by StefanPrintezis** [docker-library/php#865]
Not using cached versions. 
I've tested the install on another device, same output.

Dockerfile:
```
FROM php:7.3.7-fpm
RUN apt-get update
RUN apt-get install -y --no-install-recommends \
        libfreetype6-dev \
        libjpeg62-turbo-dev \
        libpng-dev \
    && docker-php-ext-install -j$(nproc) iconv \
    && docker-php-ext-configure gd --with-freetype-dir=/usr/include/freetype --with-jpeg-dir=/usr/include/ \
    && docker-php-ext-install -j$(nproc) gd \
    && rm -r /var/lib/apt/lists/*

```
```
configure: error: freetype-config not found.
ERROR: Service 'php' failed to build: The command '/bin/sh -c apt update && apt install -y --no-install-recommends         libfreetype6-dev         libjpeg62-turbo-dev         libpng-dev     && docker-php-ext-install -j$(nproc) iconv     && docker-php-ext-configure gd --with-freetype-dir=/usr/include/ --with-jpeg-dir=/usr/include/     && docker-php-ext-install -j$(nproc) gd     && rm -r /var/lib/apt/lists/*' returned a non-zero code: 1
```

The issue is present when using the latest version 7.3.7-fpm (buster)
The build succeeds when using 7.3.6-fpm (stretch)

TLDR solution for now
Use a tag with stretch instead of buster.

**Comment by StefanPrintezis** [comment]
This might be related switching to debian from stretch?

**Comment by jeyk** [comment]
> This might be related switching to buster from stretch?

Probably. The changelog for libfreetype6-dev states: 

- The `freetype-config' script is no longer installed by default
      (Closes: #871470, #886461). All packages depending on libfreetype6-dev
      should use pkg-config to find the relevant CFLAGS and libraries.

I guess that `docker-php-ext-configure gd` didn't anticipate this change...

**Comment by StefanPrintezis** [comment]
So somewhere this should be added?: 
```
pkg-config freetype2 --libs
pkg-config freetype2 --cflags
```

Not sure where yet.

**Comment by marcingajda** [comment]
For people looking here for a quick fix: in your Dockerfile replace this
```
FROM php:7.3-fpm
```
with this
```
FROM php:7.3-fpm-stretch
```

**Comment by rjocoleman** [comment]
This shouldn't be an issue with PHP 7.4 because of the migration to `pkg-config`. https://github.com/php/php-src/blob/PHP-7.4/UPGRADING#L606

This has been a problem for a wee while for linux distros like Arch that have newer versions running. 
https://bugs.php.net/bug.php?id=76324

The patch here adds dection via pkg-config https://git.archlinux.org/svntogit/packages.git/tree/trunk/freetype.patch?h=packages/php
(`pkg-config freetype2 --cflags` and `pkg-config freetype2 --libs` as @StefanPrintezis notes)

Extracting the PHP source early in your `Dockerfile` and patching it is one way to work around this.

```Dockerfile
FROM php:7.3.7-fpm

RUN apt-get update; \
  apt-get install -y --no-install-recommends \
  libfreetype6-dev \
  libjpeg62-turbo-dev \
  libpng-dev \
  pkg-config \
  patch;

ADD https://git.archlinux.org/svntogit/packages.git/plain/trunk/freetype.patch?h=packages/php /tmp/freetype.patch
RUN docker-php-source extract; \
  cd /usr/src/php; \
  patch -p1 -i /tmp/freetype.patch; \
  rm /tmp/freetype.patch

RUN docker-php-ext-configure gd --with-gd --with-freetype-dir 
```

I added `patch` and `pkg-config` in the example for clarity, I didn't check to see if they're included in the buster image by default.

**Comment by andreasschroth** [comment]
Kind of OT, but is it really such a good idea to switch OS version within a minor release? Didn't expect a switch from Stretch to Buster from 7.3.6 to 7.3.7. Pretty unexpected.

**Comment by barnabas-szekeres** [comment]
The issue also exist in `php:7.2-apache` image :(

**Comment by tvanro** [comment]
Same issue with `wordpress:5.2.2-apache` (that's building FROM `php:7.3-apache`)

```
FROM wordpress:5.2.2-apache
RUN apt-get update && apt-get install -y \
		libfreetype6-dev \
		libjpeg62-turbo-dev \
		libpng-dev \
	&& docker-php-ext-install -j$(nproc) iconv \
	&& docker-php-ext-configure gd --with-freetype-dir=/usr/include/ --with-jpeg-dir=/usr/include/ \
	&& docker-php-ext-install -j$(nproc) gd
```

```
configure: error: freetype-config not found.
ERROR: Service 'wordpress_apache' failed to build: The command '/bin/sh -c apt-get update && apt-get install -y 		libfreetype6-dev 		libjpeg62-turbo-dev 		libpng-dev 	&& docker-php-ext-install -j$(nproc) iconv 	&& docker-php-ext-configure gd --with-freetype-dir=/usr/include/ --with-jpeg-dir=/usr/include/ 	&& docker-php-ext-install -j$(nproc) gd' returned a non-zero code: 1
```

**Comment by euphoriatek** [comment]
Same issue on php:7.2-fpm image

image: php:7.2-fpm

this line is failing: - docker-php-ext-configure gd --with-freetype-dir=/usr/include/ --with-jpeg-dir=/usr/include/

it started happening recently.

**Comment by jxlwqq** [comment]
Temporary use: 
```
php:7.2.19-apache
```

**Comment by andreasschroth** [comment]
> Temporary use:
> 
> ```dockerfile
> FROM php:7.2.19-apache
> ```
> 
> or:
> 
> ```dockerfile
> FROM php:7.3.6-apache
> ```

No, not really a good idea. You'll end up stock with 7.2.19 or 7.3.6 and won't receive any security fixes, etc. anymore.

Better use this:

```dockerfile
FROM php 7.2-apache-stretch
```

or:

```dockerfile
FROM php 7.3-apache-stretch
```

**Comment by andreasschroth** [comment]
> Same issue on php:7.2-fpm image
> 
> image: php:7.2-fpm
> 
> this line is failing: - docker-php-ext-configure gd --with-freetype-dir=/usr/include/ --with-jpeg-dir=/usr/include/
> 
> it started happening recently.

Use a base image based on Debian Stretch instead of Buster.

```dockerfile
FROM php 7.2-fpm-stretch
```

**Comment by euphoriatek** [comment]
> 
> 
> Temporary use:
> 
> ```dockerfile
> FROM php:7.2.19-apache
> ```
> 
> or:
> 
> ```dockerfile
> FROM php:7.3.6-apache
> ```

tried using both and also php:7.2-fpm-stretch

still the same error 

checking for FreeType 2... /usr/include/
checking whether to enable JIS-mapped Japanese font support in GD... no
If configure fails try --with-webp-dir=<DIR>
configure: error: jpeglib.h not found.

**Comment by andreasschroth** [comment]
I believe most people here don't really understand what's happening / going on.

The latest released docker images with the default tags (e.g. `php:7.3-fpm`)  are now based on Debian Buster instead of Debian Stretch before.

Debian Buster uses a more up-to-date version of the `libfreetype6-dev` (`libfreetype6`) package (version `2.9.1-3` instead of `2.6.3-3.2` before). But Version 2.9.1-3 isn't shipped with `freetype-config` anymore, as it was removed and the package maintainers recommend to replace calls to `freetype-config` by calls to `pkg-config`.

So as I have already commented a few times before here: Switch to the tags suffixing with `-stretch`. E.g. instead of using `php:7.3-fpm` you uses `php:7.3-fpm-stretch`. Or instead of `php:7.2-apache` you use `php:7.2-apache-stretch`. That's it.

**Comment by jvanoosterom** [comment]
same issue with php:7.1-apache

**Comment by euphoriatek** [comment]
> 
> 
> I believe most people here don't really understand what's happening / going on.
> 
> The latest released docker images with the default tags (e.g. `php:7.3-fpm`) are now based on Debian Buster instead of Debian Stretch before.
> 
> Debian Buster uses a more up-to-date version of the `libfreetype6-dev` (`libfreetype6`) package (version `2.9.1-3` instead of `2.6.3-3.2` before). But Version 2.9.1-3 isn't shipped with `freetype-config` anymore, as it was removed and the package maintainers recommend to replace calls to `freetype-config` by calls to `pkg-config`.
> 
> So as I have already commented a few times before here: Switch to the tags suffixing with `-stretch`. E.g. instead of using `php:7.3-fpm` you use `php:7.3-fpm-stretch`. Or instead of `php:7.2-apache` you use `php:7.2-apache-stretch`. That's it.

for those using 7.2-fpm image using 7.2-fpm-stretch not worked but instead 7.2.20-fpm-stretch worked

**Comment by euphoriatek** [comment]
> 
> 
> I believe most people here don't really understand what's happening / going on.
> 
> The latest released docker images with the default tags (e.g. `php:7.3-fpm`) are now based on Debian Buster instead of Debian Stretch before.
> 
> Debian Buster uses a more up-to-date version of the `libfreetype6-dev` (`libfreetype6`) package (version `2.9.1-3` instead of `2.6.3-3.2` before). But Version 2.9.1-3 isn't shipped with `freetype-config` anymore, as it was removed and the package maintainers recommend to replace calls to `freetype-config` by calls to `pkg-config`.
> 
> So as I have already commented a few times before here: Switch to the tags suffixing with `-stretch`. E.g. instead of using `php:7.3-fpm` you use `php:7.3-fpm-stretch`. Or instead of `php:7.2-apache` you use `php:7.2-apache-stretch`. That's it.



> 
> 
> > Same issue on php:7.2-fpm image
> > image: php:7.2-fpm
> > this line is failing: - docker-php-ext-configure gd --with-freetype-dir=/usr/include/ --with-jpeg-dir=/usr/include/
> > it started happening recently.
> 
> Use a base image based on Debian Stretch instead of Buster:
> 
> ```dockerfile
> FROM php 7.2-fpm-stretch
> ```

thanks, 7.2.20-fpm-stretch worked

**Comment by goetas** [comment]
same issue with `php:7.1-fpm`, solved with changing version to  `php:7.1-fpm-stretch` (as @jvanoosterom did)

**Comment by juuuuuu** [comment]
Same here, `php:7.1-apache-stretch` FTW!

**Comment by moqmar** [comment]
The same issue was closed a year ago in #608 for not being reproducable.

Relevant issue in Debian: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=870618

**Comment by goetas** [comment]
This can be reproduced running just 
```bash
docker run --rm php:7.1.30 bash -c 'apt-get update -qq && apt-get install libfreetype6-dev -y -qq && docker-php-ext-configure gd --with-freetype-dir=/usr/include/ --with-jpeg-dir=/usr/include/'
```

**Comment by tianon** [comment]
This is a known issue in PHP upstream: https://bugs.php.net/bug.php?id=76324

For PHP 7.4+ (still a pre-release), the solution is to simply use `--with-freetype` and it will use `pkg-config` as appropriate to find `freetype`.

For older PHP releases, the solution is more complicated and likely involves some variation of https://git.archlinux.org/svntogit/packages.git/tree/php/trunk/freetype.patch, which IMO should really be applied as a backport upstream for handling `pkg-config` as a fallback in conjunction with `freetype-config`.  We might need to consider applying something like that to our builds, but that's a solution of last resort -- we provide a stock PHP upstream experience as much as possible, and applying patches that ought to be applied upstream is something we try to strongly avoid.

For many users, the solution is likely going to be switching to explicitly use `-stretch` as described above.  To be clear, this is exactly what those Debian release named aliases were added for, and why the `stretch` variants are not deprecated/removed yet.

**Comment by gries** [comment]
I'm curious on why the switch to buster was done "silently" by that I mean why isn't strech still the default and `-buster` an option? it just confused me as php at least since 7.* times is all about bc compatibility and this somewhat goes against that philosophy.

**Comment by wglambert** [comment]
Stretch isn't the default because Buster is now the ["stable"](https://wiki.debian.org/DebianStable) version of Debian which is considered the "official release"
https://wiki.debian.org/DebianStability

There isn't a particularly great way of notifying users of changes outside of them checking the [commits](https://github.com/docker-library/php/commits/master) or [pull requests](https://github.com/docker-library/official-images/labels/library%2Fphp)

Using explicit tags would help to prevent breakage from an update since `php-7.3-fpm-stretch` will resolve to the same base OS even when the default tag changes to a different one, specifying the OS variant is along the same lines as specifying version numbers. Just as you might not want to use the tag `php:7.3-fpm-alpine` as you'll be updated to any major OS changes, but instead `php:7.3-fpm-alpine3.10` which will just provide minor updates to 3.10.

Going a step further if a breakage is of particular concern then using the image's [sha256](https://github.com/docker-library/repo-info/blob/master/repos/php/remote/7.3.7-fpm-stretch.md) digest will lock you to that specific version in time, so that you could then test updates for utmost compatibility.

For a full list of tags see https://github.com/docker-library/docs/blob/master/php/README.md#supported-tags-and-respective-dockerfile-links

**Comment by regiszanandrea** [comment]
same issue with php:7.2-fpm, solved with changing version to php:7.2-fpm-stretch (as @jvanoosterom did and @goetas)

**Comment by gries** [comment]
Thanks for the clarification of the debian part :)

**Comment by DuckThom** [comment]
For now I am using the following workaround in my custom image as I'm running Magento 2.3.2 and that requires a libsodium version that's not shipped in stretch.

```
FROM php:7.2-apache

# Add the stretch repo so the old freetype6 versions are available
RUN echo "deb http://deb.debian.org/debian stretch main" >> /etc/apt/sources.list

# Force apt to install the old freetype6 version
RUN apt-get update && apt-get install -y --no-install-recommends \
    libfreetype6-dev=2.6.3-3.2 libfreetype6=2.6.3-3.2

RUN docker-php-ext-configure gd --with-freetype-dir=/usr/lib/x86_64-linux-gnu/ \
```

**Comment by cnlyzy** [comment]
`FROM php:fpm`

also have this probrem, change to 

`FROM php:fpm-stretch`

It Works for me and the error is gone.

**Comment by ManukMinasyan** [comment]
Hey guys, I have the same issue on windows 10(Docker desktop).
How I can fix it?
Thanks)

`Building horizon
Step 1/9 : FROM php:7.3-fpm
 ---> 507c954e3d50
Step 2/9 : RUN apt-get update &&     apt-get install -y --no-install-recommends         curl         libz-dev         libpq-dev         libicu-dev         libpng-dev         libjpeg-dev         libfreetype6-dev     && apt-get clean
 ---> Using cache
 ---> e099c5be13b9
Step 3/9 : RUN docker-php-ext-install intl pcntl pgsql pdo_pgsql
 ---> Using cache
 ---> d4e5ece6468d
Step 4/9 : RUN pecl install -o -f redis && docker-php-ext-enable redis
 ---> Using cache
 ---> 4cfaec247f5a
Step 5/9 : RUN docker-php-ext-configure gd         --enable-gd-native-ttf         --with-jpeg-dir=/usr/lib         --with-freetype-dir=/usr/include/freetype2 &&     docker-php-ext-install gd
 ---> Running in 57ae86013329
Configuring for:
PHP Api Version:         20180731
Zend Module Api No:      20180731
Zend Extension Api No:   320180731
configure: WARNING: unrecognized options: --enable-gd-native-ttf
checking for grep that handles long lines and -e... /bin/grep
checking for egrep... /bin/grep -E
checking for a sed that does not truncate output... /bin/sed
checking for cc... cc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables...
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C compiler... yes
checking whether cc accepts -g... yes
checking for cc option to accept ISO C89... none needed
checking how to run the C preprocessor... cc -E
checking for icc... no
checking for suncc... no
checking whether cc understands -c and -o together... yes
checking for system library directory... lib
checking if compiler supports -R... no
checking if compiler supports -Wl,-rpath,... yes
checking build system type... x86_64-pc-linux-gnu
checking host system type... x86_64-pc-linux-gnu
checking target system type... x86_64-pc-linux-gnu
checking for PHP prefix... /usr/local
checking for PHP includes... -I/usr/local/include/php -I/usr/local/include/php/main -I/usr/local/include/php/TSRM -I/usr/local/include/php/Zend -I/usr/local/include/php/ext -I/usr/local/include/php/ext/date/lib
checking for PHP extension directory... /usr/local/lib/php/extensions/no-debug-non-zts-20180731
checking for PHP installed headers prefix... /usr/local/include/php
checking if debug is enabled... no
checking if zts is enabled... no
checking for re2c... re2c
checking for re2c version... 1.1.1 (ok)
checking for gawk... no
checking for nawk... nawk
checking if nawk is broken... no
checking for GD support... yes, shared
checking for the location of libwebp... no
checking for the location of libjpeg... /usr/lib
checking for the location of libpng... no
checking for the location of libz... no
checking for the location of libXpm... no
checking for FreeType 2... /usr/include/freetype2
checking whether to enable JIS-mapped Japanese font support in GD... no
If configure fails try --with-webp-dir=<DIR>
checking for jpeg_read_header in -ljpeg... yes
checking for png_write_image in -lpng... yes
If configure fails try --with-xpm-dir=<DIR>
configure: error: freetype-config not found.
ERROR: Service 'horizon' failed to build: The command '/bin/sh -c docker-php-ext-configure gd         --enable-gd-native-ttf         --with-jpeg-dir=/usr/lib         --with-freetype-dir=/usr/include/freetype2 &&     docker-php-ext-install gd' returned a non-zero code: 1`

**Comment by MichaelBrenden** [comment]
# Freetype 2.10.1 (and maybe earlier) offers a configure option to compile its much-needed freetype-config:

cd /usr/local/src/freetype-2.10.1
./configure --prefix=/usr/local/freetype-2.10.1 --enable-freetype-config

# Then tell php configure about it

cd /usr/local/src/php-7.3.12
./configure <lots of other options> --with-freetype-dir=/usr/local/freetype-2.10.1

**Comment by perecedero** [comment]
For those still using the freetype.patch here is a working URL.
https://git.archlinux.org/svntogit/packages.git/plain/trunk/freetype.patch?h=packages/php&id=a8e8e87f405e0631b2a4552656413735ebf9457c

**Comment by JulianKingman** [comment]
I ran into this same issue, but now the automatic upgrade from buster to bullseye

Posted about my issue here: https://stackoverflow.com/questions/71443918/e-unable-to-locate-package-sqlite-docker/71443919#71443919

I found this issue flummoxing, I wouldn't expect the dependencies of a docker image to update. Is there some way to be aware of such changes?

**Comment by yosifkit** [comment]
PR's here and [`library/php` PR's in official-images](https://github.com/docker-library/official-images/pulls?q=is%3Aopen+is%3Apr+label%3Alibrary%2Fphp) are the only signals we have for letting users know of changes. The buster -> bullseye move was almost 7 months ago (https://github.com/docker-library/php/pull/1190). Tags for an official-image roughly reflect how much change the user is willing to accept on a `docker pull`. Besides for Dockerfile changes, the images will also be rebuilt for OS updates about every 30 days (see also https://github.com/docker-library/python/issues/699#issuecomment-1040583314).

 - `X-fpm` -> larger PHP updates are fine
 - `X.Y-fpm` -> base image changes are fine
 - `X.Y-fpm-bullseye` -> need a specific Debian suite 
   - only two suites are maintained at a time, so once `bullseye` was added, `stretch` was no longer updated
 - `X.Y.Z-fpm-bullseye` -> need an exact PHP version (these ones are not updated once Z+1 is released)
 - `X.Y.Z-fpm-bullseye@sha256[replace-with-sum]` -> this image will never change; the sha256 is a content addressable digest

**Comment by fhuitelec** [comment]
For wanderers like myself, here's an adaptation of https://github.com/docker-library/php/issues/865#issuecomment-511163936 with the right patch file:

```dockerfile
ADD https://gitlab.archlinux.org/archlinux/packaging/packages/php/-/raw/17ac5054f82b54aa35d98ba6e93c2a4a8c0b302e/free /tmp/freetype.patch
RUN docker-php-source extract; \
  cd /usr/src/php; \
  patch -p1 -i /tmp/freetype.patch; \
  rm /tmp/freetype.patch
```

> [!WARNING]
> The patch in [Arch Linux' package repository](https://gitlab.archlinux.org/archlinux/packaging/packages/php/-/blob/17ac5054f82b54aa35d98ba6e93c2a4a8c0b302e/freetype.patch) was created for PHP v7.2.5, you might need to recreate it yourself using the [original file](https://github.com/php/php-src/blob/PHP-7.2.5/ext/gd/config.m4#L186-L211).

> [!IMPORTANT]
> I would recommend to not rely on the remote and host the patch file yourself next to the Dockerfile: as you can see, the patch has been moved around a lot and there are no guarantee it will be hosted there forever and with the same commit hash.
